[vulkan] Shaders caching

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #41435 WIP - vulkan view
* #41434 [vulkan] cat op (concatenate)
* #41380 [vulkan] inplace add_, relu_
* #41379 [vulkan] max_pool2d
* #41224 [vulkan][asan] Fix Invalid Memory ops
* #41223 [vulkan] reshape op
* #41222 [vulkan] support add for dim < 4
* #41221 [vulkan] mm op through addmm
* #41220 [vulkan] adaptive_avg_pool2d
* **#39384 [vulkan] Shaders caching**

Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

IvanKobzarev committed:
[vulkan] Shaders caching

[ghstack-poisoned]

IvanKobzarev committed:
rebase on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/Vulkan.cpp:
```diff
@@ -896,6 +908,79 @@ VBuffer makeUniformConstBuffer(void* ptr, VkDeviceSize size) {
   return constBuffer;
 }
 
+ComputeUnitFactory::ComputeUnitFactory(const VkDevice& device)
```
Typically opaque handles in Vulkan like VkDevice etc. are just a typedef to an integer or a pointer so there's no need to pass them by reference.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/Vulkan.h:
```diff
@@ -451,6 +463,40 @@ class ComputeUnit final {
 
+class ComputeUnitFactory {
+ public:
+  ComputeUnitFactory(const VkDevice& device);
```
It's generally considered good practice in C++ to mark single argument constructors as explicit to avoid implicit type conversions unless such behavior (i.e. in this case implicit conversion of VkDevice to ComputeUnitFactory) is desired.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/Vulkan.cpp:
```diff
@@ -896,6 +908,79 @@ VBuffer makeUniformConstBuffer(void* ptr, VkDeviceSize size) {
+    WorkGroupSize& workGroupSize) {
+  std::stringstream ss;
+  ss << key << ':' << workGroupSize.x << ':' << workGroupSize.y << ':'
+     << workGroupSize.z;
```
This kind of looks like an expensive operation to construct a key for lookups if those lookups are frequent but on the other hand it may never end up mattering so I think it's fine to roll with it for now.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/Vulkan.h:
```diff
@@ -451,6 +463,40 @@ class ComputeUnit final {
+
+  VkDevice device_;
+  VkPipelineCache pipelineCache_;
+  std::map<std::string, std::shared_ptr<ComputeUnit>> computeUnits_;
```
Consider using a std::unordered_map (aka hash map) instead of a std::map.  std::map is a poor data structure in general for performance: O(log(n)) look up and poor cache performance due to scattered memory accesses which is even more detrimental to performance when dealing with few (say hundreds) of elements like we are here.  In those scenarios the best data structure is typically a vector with a good old boring linear search or for slightly bigger arrays (1000s of elements) sorted vector plus a binary search.  A hash map is perfectly fine for now, but just as an FYI if you wanted to go down the route of vector you would store [uint64_t, std::shared_ptr<ComputeUnit>] in the vector where uint64_t is the hash key, and you would do the lookup by using hash(string) as key. Collisions should be very rare and could be handled by storing identical keys inline and doing a linear search over them.

Chances are high this never ends up mattering in our usage here so unordered_map is perfectly fine.

IvanKobzarev replied:
Thanks, I did not know the difference between std::map and std::unordered_map, looks like map works as HashTree (balanced binary tree) while unordered_map juts as a hash map without preserving ordering. I will change it to unordered_map.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/Vulkan.cpp:
```diff
@@ -896,6 +908,79 @@ VBuffer makeUniformConstBuffer(void* ptr, VkDeviceSize size) {
+  return *(computeUnit.get());
+}
+
+#ifdef USE_VULKAN_SHADERC_RUNTIME
```
One alternate design could have been to encapsulate the difference between source code (null-terminated const char* shader_source) and binary representation ([const uint32_t* code,  const uint32_t size] pair) such that the need for multiple code paths could have been avoided.  Haven't put much thought into it but I think it should be possible to cheaply abstract that away.

IvanKobzarev committed:
Update on "[vulkan] Shaders caching"


Introducing `ComputeUnitFactory` which is responsible for providing `ComputeUnit`s (Shaders), 
it caches it, using shader name (glsl file name)+workGroupSize as a cacheKey, just `std::map<string, std::shared_ptr>`

Macro GLSL_SPV changed to have literal name for cache key as a first argument.

All constructors of ComputeUnit are changed to use `ComputeUnitFactory`


Ownership model:
ComputeUnitFactory also owns `vkPipelineCache` that is internal vulkan cache object ( https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineCache.html )

`VContext` (global object) owns ComputeUnitFactory, that owns ComputeUnits, vkPipelineCache, for destruction of them we need valid VkDevice, so it should be destructed before `vkDestryDevice` in `~VContext` => As members of the class will be destructed only after destructor - forcing destruction of ComputeUnitFactory before `vkDestroyDevice`, doing `unique_ptr<ComputeUnitFactory>.reset()`

Differential Revision: [D21962430](https://our.internmc.facebook.com/intern/diff/D21962430)

[ghstack-poisoned]
